hmachine/src/main/java/org/hortonmachine/hmachine/modules/statistics/kerneldensity/OmsStructuralSimilarity.java

```diff
@@ -0,0 +1,387 @@
+
+
+        int cols = inRaster1.getCols();
+        int rows = inRaster1.getRows();
+
+        KernelFactory.ValueType type = getKernelType();
+        KernelJAI kernel = KernelFactory.createCircle(pRadius, type);
+
+        HMRaster outputRaster = new HMRaster.HMRasterWritableBuilder().setTemplate(inMap1).build();
+
+        float[] kernelData = kernel.getKernelData();
+
+        pm.beginTask("Estimating kernel density...", cols - 2 * pRadius);
+
+        IntStream.range(pRadius, rows - pRadius).parallel().forEach(r -> {
+            if(errorOccurred) {
+                return;
+            }
+            for( int c = pRadius; c < cols - pRadius; c++ ) {
+
+                double inputValue1 = inRaster1.getValue(c, r);
+                if (inRaster1.isNovalue(inputValue1)) {
+                    continue;
+                }
+                double inputValue2 = inRaster2.getValue(c, r);
+                if (inRaster2.isNovalue(inputValue2)) {
+                    continue;
+                }
+
+                double meanValue1 = meanRaster1.getValue(c,r);
+                double meanValue2 = meanRaster2.getValue(c,r);
+                if (meanRaster1.isNovalue(meanValue1) || meanRaster2.isNovalue(meanValue2)) {
+                    continue;
+                }
+
+                if (doConstant) {
+                    inputValue1 = 1.0;
+                    inputValue2 = 1.0;
+                }
+                int k = 0;
+                double outputValue = 0.0;
+                for( int kr = -pRadius; kr <= pRadius; kr++ ) {
+                    for( int kc = -pRadius; kc <= pRadius; kc++ ) {
+                        double value1 = inRaster1.getValue(c + kc, r + kr);
+                        if (inRaster1.isNovalue(value1)) {
+                            value1 = 0;
+                        }
+                        double value2 = inRaster2.getValue(c + kc, r + kr);
+                        if (inRaster2.isNovalue(value2)) {
+                            value2 = 0;
+                        }
+                        try {
+                            outputValue = outputValue + kernelData[k++] * (value1 - meanValue1) * (value2 - meanValue2);
+                        } catch (Exception e) {
+                            throw new RuntimeException(e);
+                        }
+                    }
+                }
+                try {
+                    outputRaster.setValue(c, r, outputValue);
+                } catch (IOException e) {
+                    errorOccurred = true;
+                    errorMessage = e.getLocalizedMessage();
+                }
+            }
+            pm.worked(1);
+        });
+        pm.done();
+
+        if (errorOccurred) {
+            throw new ModelsRuntimeException(errorMessage, this);
+        }
+
+        return HMRaster.fromGridCoverage(outputRaster.buildCoverage());
```

Comment:
Here just return outputRaster, it is a HMRaster already.

Reply:
That's fixed in the new commit.